doc: extensions: doxygen: add support for multiple projects

gmarull opened the pull request:
Add support for multiple Doxygen projects to doxyrunner/doxybridge extensions. While not needed upstream, some other projects may benefit from this option when re-using the extension.

Also refactored the api_overview extension as it had some major flaws (see commit for details)

gmarull committed:
doc: extensions: api_overview: refactor extension

The extension had a some major design flaws, mainly:

- Ignored the `api_overview_doxygen_xml_dir` setting, instead it
  "sniffed" doxyrunner properties, so violating environment boundaries
- Computation of Doxygen HTML path worked because of the hardcoded URL
  in relative form found in doc/conf.py

This patch moves most code to the actual directive, so that context can
be obtained from the document being parsed. Also, the only config
required now is the Doxygen output dir, obtained from doxyrunner at
conf.py level.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
doc: extensions: doxyrunner: add support for multiple projects

Modify the extension so that multiple Doxygen projects can be added.
While not required in upstream Zephyr, other users downstream may
require support for this feature.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
doc: extensions: doxybridge: add support for multiple projects

Modify the extension so that multiple Doxygen projects can be
referenced. While not required in upstream Zephyr, other users
downstream may require support for this feature.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

kartben commented on doc/_extensions/zephyr/api_overview.py:
```diff
@@ -21,156 +43,121 @@ class ApiOverview(SphinxDirective):
-        ][0]
-    except IndexError as e:
-        raise Exception(f"Unexpected group {innergroup.get_refid()}") from e
+            for g in groups
+            if g.get_compounddef()[0].get_id()
+            not in [
+                i.get_refid()
+                for h in [j.get_compounddef()[0].get_innergroup() for j in groups]
+                for i in h
+            ]
```
It's quite a mouthful. Could use a comment, or slightly more explicit variable names?

gmarull replied:
I just fixed the most important issues in the extension, but not any further issues like unreadable code. I think this should be done by the original authors.

gmarull replied:
I have actually added a commit improving this, please check

kartben commented on doc/_extensions/zephyr/api_overview.py:
```diff
@@ -21,156 +43,121 @@ class ApiOverview(SphinxDirective):
-    rows.append(row_node)
+    def visit_group(self, group, all_groups, rows, indent=0):
+        version = since = ""
+        github_uri = "https://github.com/zephyrproject-rtos/zephyr/releases/tag/"
```
Not sure we want to hardcode this?

gmarull replied:
It was already hardcoded, but I will add a commit to move this to an option.

gmarull replied:
fixed

gmarull committed:
doc: doc-guidelines: document :project: option for doxygengroup

The doxygengroup now supports specifying a project, document this
feature.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
doc: extensions: api_overview: improve some comprehensive lists

Some comprehensive lists were hard to read, improve the code a bit so
that things make more sense.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
doc: extensions: api_overview: do not hardcode github URL

Add a new extension config value so that repo URL is not hardcoded in
the extension code.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull commented on doc/_extensions/zephyr/doxybridge.py:
```diff
@@ -34,6 +35,9 @@ class DoxygenGroupDirective(SphinxDirective):
     required_arguments = 1
     optional_arguments = 0
+    option_spec = {
+        "project": directives.unchanged,
```
done

kartben commented on doc/contribute/documentation/guidelines.rst:
```diff
@@ -1006,6 +1006,13 @@ Doxygen API documentation
 
       .. doxygengroup:: can_interface
 
+
+   Doxygen project can optionally be specified using the ``:project:`` option.
+   This can be useful when multiple Doxygen projects are configured. Example::
+
+      .. doxygengroup:: can_interface
+         :project: zephyr
+
```
do you mind using a construct similar to https://github.com/zephyrproject-rtos/zephyr/blob/main/doc/contribute/documentation/guidelines.rst?plain=1#L1086-L1091?
Beyond the arguably more consistent formatting, this also gives us an HTML anchor "for free" that allows to directly link to the doc of that particular option should one wish to. Thanks!

gmarull replied:
done
